wagtail/contrib/modeladmin/helpers/search.py

```diff
@@ -0,0 +1,61 @@
+import operator
+from functools import reduce
+
+from django.contrib.admin.utils import lookup_needs_distinct
+from django.db.models import Q
+
+from wagtail.search.backends import get_search_backend
+
+
+class BaseSearchHandler:
+    def __init__(self, search_fields):
+        self.search_fields = search_fields
+
+    def search_queryset(self, queryset, search_term, **kwargs):
+        """
+        Returns a tuple containing a queryset to implement the search,
+        and a boolean indicating if the results may contain duplicates.
+        """
+        raise NotImplementedError()
+
+    @property
+    def show_search_form(self):
+        """
+        Returns a boolean that determines whether a search form should be
+        displayed in the IndexView UI
+        """
+        return True
+
+
+class DjangoORMSearchHandler(BaseSearchHandler):
+    def search_queryset(self, queryset, search_term):
+        if not search_term or not self.search_fields:
+            return queryset, False
+        use_distinct = False
+        orm_lookups = ['%s__icontains' % str(search_field)
+                       for search_field in self.search_fields]
+        for bit in search_term.split():
+            or_queries = [Q(**{orm_lookup: bit})
+                          for orm_lookup in orm_lookups]
+            queryset = queryset.filter(reduce(operator.or_, or_queries))
+        opts = queryset.model._meta
+        for search_spec in orm_lookups:
+            # Check wether out results may have duplicates, then remove them
+            if lookup_needs_distinct(opts, search_spec):
+                use_distinct = True
+                break
+        return queryset, use_distinct
+
+    @property
+    def show_search_form(self):
+        return bool(self.search_fields)
+
+
+class WagtailBackendSearchHandler(BaseSearchHandler):
+    def search_queryset(self, queryset, search_term, backend='default'):
```

Comment:
Hi @seb-b - I like that each handler can accept different arguments, but I don't see a way in the current design for the `IndexView` to easily make use of this. I can't help but think additional methods/attributes on the `ModelAdmin` class would be a good way to facilitate this, as it would make things easier to override (you wouldn't have to create a custom `IndexView` class just to change the behaviour.

A common pattern for this sort of thing is to add an attribute that can be overridden to change the class-wide behaviour (e.g. `search_handler_extra_search_kwargs`), but accompany it with a method to allow more contextual overriding where needed (e.g. `get_search_handler_extra_search_kwargs(self, request, queryset, search_term)`.

Also, I see no reason why `WagtailBackendSearchHandler.search_queryset()` here shouldn't accept all of the arguments accepted by [`SearchableQuerysetMixin.search()`](https://github.com/wagtail/wagtail/blob/master/wagtail/search/queryset.py#L5)

Does this sound good to you?